pkg/apis/stackconfigpolicy/v1alpha1/stackconfigpolicy_types.go

```diff
@@ -0,0 +1,210 @@
+}*/
+
+type StackConfigPolicyStatus struct {
+	// ResourcesStatuses holds the status for each resource to be configured.
+	ResourcesStatuses map[string]ResourcePolicyStatus `json:"resourcesStatuses"`
+	// Resources is the number of resources to be configured.
+	Resources int `json:"resources,omitempty"`
+	// Ready is the number of resources successfully configured.
+	Ready int `json:"ready,omitempty"`
+	// Errors is the number of resources which have an incorrect configuration
+	Error int `json:"errors,omitempty"`
+	// ReadyCount is a Human representation of the number of resources successfully configured.
+	ReadyCount string `json:"readyCount,omitempty"`
+	// Phase is the phase of the StackConfigPolicy.
+	Phase PolicyPhase `json:"phase,omitempty"`
+	// ObservedGeneration is the most recent generation observed for this StackConfigPolicy.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+type PolicyPhase string
+
+const (
+	UnknownPhase         PolicyPhase = "Unknown"
+	ReadyPhase           PolicyPhase = "Ready"
+	ApplyingChangesPhase PolicyPhase = "ApplyingChanges"
+	InvalidPhase         PolicyPhase = "Invalid"
+	ErrorPhase           PolicyPhase = "Error"
+	ConflictPhase        PolicyPhase = "Conflict"
+)
+
+// phaseOrder
+var phaseOrder = map[PolicyPhase]int{
+	UnknownPhase:         -1,
+	ReadyPhase:           0,
+	ApplyingChangesPhase: 1,
+	InvalidPhase:         2,
+	ErrorPhase:           3,
+	ConflictPhase:        4,
+}
+
+// ResourcePolicyStatus models the status of the policy for one resource to be configured.
+type ResourcePolicyStatus struct {
+	Phase           PolicyPhase       `json:"phase,omitempty"`
+	CurrentVersion  int64             `json:"currentVersion,omitempty"`
+	ExpectedVersion int64             `json:"expectedVersion,omitempty"`
+	Error           PolicyStatusError `json:"error,omitempty"`
+}
+
+type PolicyStatusError struct {
+	Version int64    `json:"version,omitempty"`
+	Errors  []string `json:"errors,omitempty"`
+}
+
+func NewStatus(scp StackConfigPolicy) StackConfigPolicyStatus {
+	status := StackConfigPolicyStatus{
+		ResourcesStatuses:  map[string]ResourcePolicyStatus{},
+		Phase:              ReadyPhase,
+		ObservedGeneration: scp.Generation,
+	}
+	status.setReadyCount()
+	return status
+}
+
+func (s *StackConfigPolicyStatus) setReadyCount() {
+	s.ReadyCount = fmt.Sprintf("%d/%d", s.Ready, s.Resources)
+}
+
+func (s *StackConfigPolicyStatus) AddPolicyErrorFor(resource types.NamespacedName, phase PolicyPhase, msg string) {
+	s.ResourcesStatuses[resource.String()] = ResourcePolicyStatus{
+		Phase: phase,
+		Error: PolicyStatusError{Errors: []string{msg}},
+	}
+	s.Update()
+}
+
+func (s *StackConfigPolicyStatus) UpdateResourceStatusPhase(resource types.NamespacedName, status ResourcePolicyStatus) {
+	if status.CurrentVersion == unknownVersion { //nolint:gocritic
+		status.Phase = UnknownPhase
+	} else if status.Error.Errors != nil {
+		status.Phase = ErrorPhase
```

Comment:
During my tests it feels a bit frustrating that the `Phase` was stuck in `Error` while I submitted a "fixed" resource:

```yaml
      currentVersion: 1668764271475691000
      error:
        errors:
        - |-
          Error processing cluster_settings state change: java.lang.IllegalArgumentException: failed to parse setting [indices.recovery.max_bytes_per_sec] with value [XXXmb]
          Caused by: org.elasticsearch.ElasticsearchParseException: failed to parse setting [indices.recovery.max_bytes_per_sec] with value [XXXmb]
          Caused by: java.lang.NumberFormatException: For input string: "xxx"
        version: 1668764417220068000
      expectedVersion: 1668764572205542000
      phase: Error
```

In that situation the `expectedVersion` (the fixed resource) is greater than `version`, would it be a reason to switch back to `ApplyingChangesPhase`? 

As the a side note the output seems a bit odd (`error` field which holds an `errors` one)

Reply:
> In that situation the expectedVersion (the fixed resource) is greater than version, would it be a reason to switch back to ApplyingChangesPhase?

Very good idea.

> As the a side note the output seems a bit odd (error field which holds an errors one)

Same feeling. I wanted to respect the format returned by Elasticsearch but it doesn't make much sense here. I think I can  replace `Errors []string` by `Message string` witout much effort. If there are several errors, I concact them in the message.